### ISSUE TITLE
[libc++] `ranges::ends_with` should forward to `equal`

### ISSUE THREAD

**Issue by philnik777** [llvm/llvm-project#83880]
`equal` forwards to `memcmp` for trivially equality comparable types, which is an order of magnitude faster than the generic loop. We should make use of that optimization in `ends_with`.

**Comment by xiaoyang-sde** [comment]
`ranges::ends_with` invokes `ranges::start_with` if the iterators model `std::bidirectional_iterator`, and invokes `ranges::equal` otherwise.

It seems that the implementation of `ranges::start_with` relies on `ranges::mismatch`, which, in turn, is implemented using a naive element-wise comparison. I believe it's possible to optimize `ranges::starts_with` using `ranges::equal`.

I'm willing to submit a patch if no one else is working on it.

**Comment by changkhothuychung** [comment]
I would like to try this if this is available.

**Comment by ldionne** [comment]
I don't think that optimizing `std::starts_with` is sufficient to optimize this though, since `ends_with` calls `starts_with` with `reverse_iterator`. We'd have to unwrap the `reverse_iterator` if it's wrapping a `contiguous_iterator` first, I think.
